util/dvsim/LsfLauncher.py

```diff
@@ -0,0 +1,379 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import logging as log
+import os
+import re
+import subprocess
+from pathlib import Path
+
+from Launcher import Launcher, LauncherError
+from utils import VERBOSE, clean_odirs
+
+
+class LsfLauncher(Launcher):
+
+    # A hidden directory specific to a cfg, where we put individual 'job'
+    # scripts.
+    jobs_dir = {}
+
+    # All launcher instances available for lookup.
+    jobs = {}
+
+    # When the job completes, we try to read the job script output to determine
+    # the outcome. It may not have been completely written the first time we
+    # read it so we retry on the next poll, no more than 10 times.
+    max_poll_retries = 10
+
+    # TODO: Add support for build/run/cov job specific resource requirements:
+    #       cpu, mem, disk, stack.
+    # TODO: Allow site-specific job resource usage setting using
+    #       `DVSIM_LSF_CFG` environment variable.
+
+    def __init__(self, deploy):
+        super().__init__(deploy)
+
+        # Set the status. Only update after the job is done - i.e. status will
+        # transition from None to P/F/K.
+        self.status = None
+
+        # Maintain the job script output as an instance variables for polling
+        # and cleanup.
+        self.job_script_out = None
+
+        # If we already opened the job script output file (but have not
+        # determined the outcome), then we maintain the file descriptor rather
+        # then reopening it and starting all over again on the next poll.
+        self.job_script_out_fd = None
+        self.job_script_out_err_msg = []
+        self.job_script_out_err_msg_found = False
+
+        # Set the job id.
+        self.job_id = None
+
+        # Polling retry counter..
+        self.num_poll_retries = 0
+
+    @staticmethod
+    def prepare_workspace_for_cfg(cfg):
+        '''This is a static method to prepare the workspace before launching
+        any job using this method, but specific to a config.
+        '''
+
+        # Create the job dir.
+        LsfLauncher.jobs_dir[cfg] = Path(cfg.scratch_path, "lsf",
+                                         cfg.timestamp)
+        clean_odirs(odir=LsfLauncher.jobs_dir[cfg], max_odirs=2)
+        os.makedirs(Path(LsfLauncher.jobs_dir[cfg]), exist_ok=True)
+
+        LsfLauncher.jobs[cfg] = {}
+
+    @staticmethod
+    def make_job_array_script_text(cfg, job_name):
```

Comment:
Oh, that's interesting! Makes sense, since presumably that's 20k NFS round-trips. I wonder whether it would be worth adding a comment somewhere near here explaining why we do things this way?